ArduCopter/Copter.cpp

```diff
@@ -445,6 +445,28 @@ bool Copter::has_ekf_failsafed() const
     return failsafe.ekf;
 }
 
+// get target location (for use by scripting)
+bool Copter::get_target_location(Location& target_loc)
+{
+    return flightmode->get_wp(target_loc);
+}
+
+/*
+  update_target_location() acts as a wrapper for set_target_location
+ */
+bool Copter::update_target_location(const Location &old_loc, const Location &new_loc)
+{
+    Location next_WP_loc;
+    flightmode->get_wp(next_WP_loc);
+    if (!old_loc.same_loc_as(next_WP_loc)) {
+        return false;
+    }
+    next_WP_loc = new_loc;
```

Comment:
Let's remove these lines stopping the update from changing the alt-frame and then we can merge.

Reply:
@rmackay9 Thank you! I fixed it.

Reply:
Ah, sorry, I wasn't clear about which two lines to remove.  The check of the old_loc should stay, the changing of the alt frame should be removed.  Sorry about that.

Reply:
By the way, we discussed on the dev call what the purpose of update_target_wp() is compared with set_target_wp() and the purpose is to avoid a race condition where some other part of the system or the user has adjusted the target location.  Apparently in these cases, the ship landing code needs to know this and handle it.

TBH, I question whether this is really necessary because Copter is stricter than plane when it comes to accepting target location, positions, velocities, etc.  In general Copter does not have issues with multiple parts of the system fighting with each other for control.

Reply:
@rmackay9 Thank you for discussing this PR. I probably understand the purpose of update_target_location. I know that copter systems are powerful. I just want to test ship landing lua script with as few changes as possible.

Reply:
> In general Copter does not have issues with multiple parts of the system fighting with each other for control.

The race condition is there no matter how good the copter code is. The lua script may be basing the new target on the existing target. If the existing target has changed then the new target will be wrong. There is no way for the copter code to know it is wrong. The update function avoids this race by rejecting the change to the target if the caller can't provide the correct existing target.